common/flatpak-image-source.c

```diff
@@ -0,0 +1,225 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authors:
+ *       Owen Taylor <otaylor@redhat.com>
+ */
+
+#include <glib/gi18n-lib.h>
+
+#include "flatpak-image-source-private.h"
+#include "flatpak-oci-registry-private.h"
+
+struct FlatpakImageSource
+{
+  GObject parent;
+
+  FlatpakOciRegistry *registry;
+  char *repository;
+  char *digest;
+
+  FlatpakOciManifest *manifest;
+  gsize manifest_size;
+  FlatpakOciImage *image_config;
+};
+
+typedef struct
+{
+  GObjectClass parent_class;
+} FlatpakImageSourceClass;
+
+G_DEFINE_TYPE (FlatpakImageSource, flatpak_image_source, G_TYPE_OBJECT)
+
+
+static void
+flatpak_image_source_finalize (GObject *object)
+{
+  FlatpakImageSource *self = FLATPAK_IMAGE_SOURCE (object);
+
+  g_clear_object (&self->registry);
+  g_free (self->repository);
+  g_free (self->digest);
+  g_clear_object (&self->manifest);
+  g_clear_object (&self->image_config);
+
+  G_OBJECT_CLASS (flatpak_image_source_parent_class)->finalize (object);
+}
+
+static void
+flatpak_image_source_class_init (FlatpakImageSourceClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = flatpak_image_source_finalize;
+}
+
+static void
+flatpak_image_source_init (FlatpakImageSource *self)
+{
+}
+
+static FlatpakImageSource *
+flatpak_image_source_new (FlatpakOciRegistry *registry,
+                          const char         *repository,
+                          const char         *digest,
+                          GCancellable       *cancellable,
+                          GError            **error)
+{
+  g_autoptr(FlatpakImageSource) self = NULL;
+  g_autoptr(FlatpakOciVersioned) versioned = NULL;
+
+  self = g_object_new (FLATPAK_TYPE_IMAGE_SOURCE, NULL);
+  self->registry = g_object_ref (registry);
+  self->repository = g_strdup (repository);
+  self->digest = g_strdup (digest);
+
+  versioned = flatpak_oci_registry_load_versioned (self->registry, self->repository,
+                                                   self->digest, NULL, &self->manifest_size,
+                                                   cancellable, error);
+  if (versioned == NULL)
+    return NULL;
+
+  if (!FLATPAK_IS_OCI_MANIFEST (versioned))
+    {
+      flatpak_fail_error (error, FLATPAK_ERROR_INVALID_DATA, _("Image is not a manifest"));
+      return NULL;
+    }
+
+  self->manifest = FLATPAK_OCI_MANIFEST (g_steal_pointer (&versioned));
```

Comment:
Is this guaranteed to be non-NULL? The old code in flatpak_mirror_image_from_oci() had an explicit check:
  if (manifest->config.digest == NULL)
    return flatpak_fail_error (error, FLATPAK_ERROR_INVALID_DATA, _("Image is not a manifest"));

Reply:
Good catch, seems like this check just got dropped - the property isn't marked mandatory in flatpak-json-oci.c so it seems like it could be NULL.

Pushed a new version with this check added back. It's not rebased because there are no conflicts - not sure what the preference is.